src/sources/soundsourceffmpeg.cpp

```diff
@@ -304,672 +416,843 @@ SoundSourceFFmpeg::~SoundSourceFFmpeg() {
+                    const auto rewindSampleLength =
+                            m_sampleBuffer.shrinkAfterWriting(
+                                    frames2samples(rewindRange.length()));
+                    rewindRange.shrinkBack(
+                            samples2frames(rewindSampleLength));
+                    // ...then rewind remaining samples from the output buffer
+                    if (pOutputSampleBuffer) {
+                        pOutputSampleBuffer -= frames2samples(rewindRange.length());
                     }
-                    return false;
+                    writableRange = IndexRange::between(rewindRange.start(), writableRange.end());
                 }
+                // Adjust read position
+                readFrameIndex = decodedFrameRange.start();
+            } else if (decodedFrameRange.start() > readFrameIndex) {
+                // The next frame starts AFTER the current position, i.e.
+                // some frames before decodedFrameRange.start() are missing.
+                missingFrameCount = decodedFrameRange.start() - readFrameIndex;
             }
 
-            // If bytes left is bigger than FFmpeg frame bytes available
-            // then copy to buffer end and then jump to next FFmpeg frame
-            // to understand this here are some examples
-            //   * MP3 have size 2304 * 4
-            //   * OGG/Opus size 256 - 1024
-            //   * WMA size 32767 - 131070
-            // and all these are separated in packets nor solid stream of bytes
-            // that just can be copied to buffer
-            // so that's why this kind of abstraction is needed
-            if (l_lLeft > (l_SObj->length - l_lOffset)) {
-                // calculate start point of copy
-                l_lBytesToCopy = l_SObj->length - l_lOffset;
-                memcpy(l_pBuffer, (l_SObj->bytes + l_lOffset), l_lBytesToCopy);
-                l_lOffset = 0;
-                l_pBuffer += l_lBytesToCopy;
-                l_lLeft -= l_lBytesToCopy;
+#if ENABLE_TRACING
+            kLogger.trace()
+                    << "Before resampling:"
+                    << "m_curFrameIndex" << m_curFrameIndex
+                    << "readFrameIndex" << readFrameIndex
+                    << "decodedFrameRange" << decodedFrameRange
+                    << "writableRange" << writableRange
+                    << "missingFrameCount" << missingFrameCount;
+#endif
+            DEBUG_ASSERT(readFrameIndex <= decodedFrameRange.start());
+            // NOTE: Decoding might start at a negative position for the first
+            // frame of the file! In this case readFrameIndex < decodedFrameRange().start(),
+            // i.e. the decoded frame starts outside of the track's valid range!
+            // Consequently isValidFrameIndex(readFrameIndex) might return false.
+            // This is expected behavior and will be compensated during 'preskip'
+            // (see below).
+
+            if (m_pSwrContext) {
```

Comment:
I would hope that resampling is in own method or own file in here it makes this too long.